src/control/cmd/daos_agent/procmon.go

```diff
@@ -135,56 +141,115 @@ func NewProcMon(logger logging.Logger) *procMon {
 	select {
 	case <-ctx.Done():
-	case p.request <- req:
+	case p.request <- request:
 	}
 }
 
-func (p *procMon) handleProcessAttach(ctx context.Context, request *procMonRequest) {
+func (p *procMon) handlePoolConnect(ctx context.Context, request *procMonRequest) {
+	p.log.Debugf("Received request to connect pool:%s with handle %s, for pid:%d", request.poolUUID, request.poolHandleUUID, request.pid)
+
 	info, found := p.procs[request.pid]
 
-	if found {
-		p.log.Errorf("Attempted to monitor process %d more than once", request.pid)
+	if !found {
+		child, cancel := context.WithCancel(ctx)
+		info = &procInfo{
+			log:       p.log,
+			pid:       request.pid,
+			cancelCtx: cancel,
+			response:  p.response,
+			handles:   make(map[string]map[string]bool),
+		}
+
+		p.procs[request.pid] = info
+		go info.monitorProcess(child)
+	}
+
+	_, found = info.handles[request.poolUUID]
+	if !found {
+		info.handles[request.poolUUID] = make(map[string]bool)
+	}
+	info.handles[request.poolUUID][request.poolHandleUUID] = true
+
+}
+
+func (p *procMon) handlePoolDisconnect(request *procMonRequest) {
+	p.log.Debugf("Received request to disconnect pool:%s with handle %s, for pid:%d", request.poolUUID, request.poolHandleUUID, request.pid)
+
+	info, found := p.procs[request.pid]
+
+	if !found || len(info.handles) == 0 {
+		p.log.Debugf("Process has no registered pools to disconnect from.")
 		return
 	}
 
-	child, cancel := context.WithCancel(ctx)
-	info = &procInfo{
-		log:       p.log,
-		pid:       request.pid,
-		cancelCtx: cancel,
-		response:  p.response,
+	if info.handles[request.poolUUID][request.poolHandleUUID] {
+		delete(info.handles[request.poolUUID], request.poolHandleUUID)
+		if len(info.handles[request.poolUUID]) == 0 {
+			delete(info.handles, request.poolUUID)
+		}
 	}
 
-	p.procs[request.pid] = info
-	go info.monitorProcess(child)
+}
+
+func (p *procMon) cleanupLeakedHandles(info *procInfo) {
+	if len(info.handles) == 0 {
+		p.log.Debugf("No leaked pool handles to clean up for pid: %d", info.pid)
+		return
+	}
+
+	for poolUUID, element := range info.handles {
+		p.log.Debugf("Cleaning up %d leaked handles from Pool UUID: %s\n", len(element), poolUUID)
+		for poolHandleUUID, _ := range element {
+			p.log.Debugf("\t%s\n", poolHandleUUID)
+		}
+		// We should construct the data structure for the gRPC call here.
+	}
+	// This is where we should make a call to the control plane with the
+	// combined set of all poolUUID and their associated poolHandleUUIDs
+
+	delete(p.procs, info.pid)
 }
 
 func (p *procMon) handleProcessDisconnect(request *procMonRequest) {
 	info, found := p.procs[request.pid]
 
 	p.log.Debugf("Received request to disconnect pid:%d\n", request.pid)
-	if !found {
-		p.log.Errorf("Attempted to disconnect process %d but process is missing", request.pid)
+	if found {
+		info.cancelCtx()
+		p.cleanupLeakedHandles(info)
+		p.log.Debugf("Process %d has disconnected", info.pid)
```

Comment:
Good debug message, but has potential to create a lot of entries.  Consider removing this as well.

Reply:
Removed.